### ISSUE TITLE
fedora 21: grive compilation fails

### ISSUE THREAD

**Issue by jakkul** [Grive/grive#290]
Standard grive cannot find json-c libraries.

Solution:

cmake/Modules/FindJSONC.cmake change lines:

FIND_PATH(JSONC_INCLUDE_DIR json-c/json.h)
FIND_LIBRARY(JSONC_LIBRARY NAMES json-c)

You also need to update the include statements in .cc files in code.

**Comment by lucquera** [comment]
everything runs ok, until 89%:

[ 89%] Building CXX object grive/CMakeFiles/grive_executable.dir/src/main.cc.o
Linking CXX executable grive
../libgrive/libgrive.a(State.cc.o): In function `gr::v1::State::Write(boost::filesystem::path const&) const':
State.cc:(.text+0x190a): undefined reference to `gr::Json::Json<long>(long const&)'
State.cc:(.text+0x199d): undefined reference to `gr::Json::Json<unsigned long>(unsigned long const&)'
State.cc:(.text+0x1a89): undefined reference to `gr::Json::Json<long>(long const&)'
collect2: error: ld returned 1 exit status
grive/CMakeFiles/grive_executable.dir/build.make:102: recipe for target 'grive/grive' failed
make[2]: *** [grive/grive] Error 1
CMakeFiles/Makefile2:162: recipe for target 'grive/CMakeFiles/grive_executable.dir/all' failed
make[1]: *** [grive/CMakeFiles/grive_executable.dir/all] Error 2
Makefile:117: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by elken** [comment]
Have you tried a full recompilation? And can you verify you actually have both json-c and json-c-devel installed?

**Comment by lucquera** [comment]
I started from the begining, and I checked, and json-c-0.12-5.fc21.i686 and json-c-devel-0.12-5.fc21.i686 are installed.

**Comment by elken** [comment]
Is the json.h file located under `/usr/include/json-c/json.h`? If it is
then everything should be working fine. :/

**Comment by lucquera** [comment]
Yes, json.h file is located in there. Tried again as root and got the same error with this State.cc.o  file. Don't know why compilation fails :/

**Comment by lucquera** [comment]
fixed with #193. It seems a problem with 32bit

**Comment by elken** [comment]
I figured it was, was going to dive through the code later but glad you
sorted :)

**Comment by rseward** [comment]
My compile breaks with a similar problem. Fedora 20 64 bit with the following json-c deps installed.

json-c-0.11-6.fc20.i686
json-c-devel-0.11-6.fc20.x86_64
json-c-0.11-6.fc20.x86_64

[ 89%] Building CXX object grive/CMakeFiles/grive_executable.dir/src/main.cc.o
Linking CXX executable grive
/bin/ld: ../libgrive/libgrive.a(Json.cc.o): undefined reference to symbol 'json_object_array_get_idx'
/bin/ld: note: 'json_object_array_get_idx' is defined in DSO /lib64/libjson-c.so.2 so try adding it to the linker command line
/lib64/libjson-c.so.2: could not read symbols: Invalid operation
collect2: error: ld returned 1 exit status
make[2]: *** [grive/grive] Error 1
make[1]: *** [grive/CMakeFiles/grive_executable.dir/all] Error 2
make: *** [all] Error 2

I used to write C++ about 10 years ago. But I am a bit lost with the slick new tools (cmake, etc.). Suggestions for fixing the linker line would be appreciated. I can retry the build on a Fedora 21 64 bit machine I have if that helps.

**Comment by lucquera** [comment]
I got this error too. #292 fixed too me. Have you tried?

**Comment by elken** [comment]
It looks like a gcc problem, gcc can't find the library for some reason.
Try adding it to CMAKE_CXX_FLAGS.

**Comment by pro-a-v** [comment]
in cmake/Modules/FindJSONC.cmake
#  JSONC_INCLUDE_DIR - The JSON-C include directory.
#  JSONC_LIBRARY     - The JSON-C library to link against.

SET(JSONC_INCLUDE_DIR /usr/include/json-c/)
SET(JSONC_LIBRARY json-c)
SET(JSONC_FOUND TRUE)

IF (JSONC_FOUND)


in /libgrive/src/protocol/Json.cc

#include "util/StdioFile.hh"

#include <json-c/json_tokener.h>   //was json
#include <json-c/linkhash.h>          //was json

#include <cassert>

in 380 str in file some rude but works:
	if ( json == 0 )
		BOOST_THROW_EXCEPTION( Error() << expt::ErrMsg( "ParseFile error" ) ) ;



DONE.
cmake
make make install
